File: sources/opencv-samples/5_ReadVideo/main.py
# https://note.nkmk.me/en/python-opencv-videocapture-file-camera/

import cv2 as cv
import argparse
import os
import signal

global has_to_break


def sigint_handler(signal, frame):
    global has_to_break
    has_to_break = True


def main():
    global has_to_break
    has_to_break = False

    args = parser.parse_args()

    if not os.path.exists(args.video_path):
        print(f"Video {os.path.exists(args.video_path)} not found")

    cap = cv.VideoCapture(args.video_path)

    if not cap.isOpened():
        print(f"Can't open video {args.video_path}")
        exit(0)

    window_name = "Read Video"

    while True:
          
        ret, frame = cap.read()
        
        if ret == False or frame is None:
            break

        if args.grayscale:
            # Our operations on the frame come here
            frame = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
          
        cv.imshow(window_name, frame)

        if cv.waitKey(1) == 27 or has_to_break:  #  or cv.getWindowProperty(window_name, cv.WND_PROP_VISIBLE) == -1
            break

    print("Video successfully played")

    # After the loop release the cap object
    cap.release()

    # Destroy all the windows
    cv.destroyAllWindows()


if __name__ == '__main__':
    signal.signal(signal.SIGINT, sigint_handler)
    parser = argparse.ArgumentParser(description="Read Video main.py CLI")
    parser.add_argument("video_path", type=str, help="path to video file")
    parser.add_argument("-g", "--grayscale", action="store_true", help="read the video with GRAYSCALE")
    main()

File: sources/opencv-samples/5_ReadVideo/main.cpp
// https://www.selfmadetechie.com/how-to-create-a-webcam-video-capture-using-opencv-c

#include <stdio.h>
#include <csignal>
#include <iostream>
#include <opencv2/opencv.hpp>
using namespace cv;
using namespace std;

bool hasToBreak = false;

void signal_handler(int signal)
{
    hasToBreak = true;
}

int main(int argc, char** argv)
{   
    // Install a signal handler
    std::signal(SIGINT, signal_handler);

    if ( argc < 2 || argc > 3 )
    {
        printf("Usage: ./main.out <video_path> [g|G]\n");
        printf("Read the README.md\n");
        return -1;
    }
    
    string video = cv::samples::findFile(argv[1]);
    VideoCapture cap(video);

    if (!cap.isOpened())
    {
        cout << "Can't open video " << video << endl;
        return 0;
    }

    bool convertToGray = false;
    if (argc == 3)
    {
        if (!strcmp("g", argv[2]) || !strcmp("G", argv[2]))
        {
            convertToGray = true;
        }
        else
        {
            printf("Unexpected parameter given\n");
            printf("Read the README.md\n");
            return -1;
        }
    }

    Mat frame;
    string window_name = "Read Video";
    namedWindow(window_name);
    
    while (1)
    {
        // get frame from the video
        cap >> frame;

        if (frame.empty()) 
        {
            break;
        }

        if (convertToGray) 
        {
            // Our operations on the frame come here
            cvtColor(frame, frame, cv::COLOR_BGR2GRAY);
        }

        imshow(window_name, frame);

        // stop the program if
        if (hasToBreak ||                                                   // due to Ctrl+C
            waitKey(1) == 27 ||                                             // ESC key
            cv::getWindowProperty(window_name, WND_PROP_AUTOSIZE) == -1)    // window is closed
        {
            break;
        }
    }
    printf("Video successfully played\n");
    cap.release();
    cv::destroyAllWindows();
    return 0;
}